Reworking keyboardShouldPersistTaps to have a middle ground

MaxLap opened the pull request:
Right now, the ScrollView's keyboard hiding behavior is either all or nothing: Hide the keyboard on any tap, or do nothing ever. This PR introduces a third mode to keyboardShouldPersistTaps which is much closer to what I consider should be the default.

In the new behavior, the tap responding is done in the bubbling phase (instead of the capture phase like =true). As a result, a child can handle the tap. If no child does, then the ScrollView will receive the tap and will hide the keyboard. As a result, changing TextInput focus works as a user expects, with a single tap and without keyboard hiding. But taping on Text or on the empty part of the ScrollView hides the keyboard and removes the focus.

You can view the behavior in a monkey patched ScrollView demo on rnplay:
https://rnplay.org/apps/E90UYw
https://rnplay.org/apps/UGzhKA

In order to have a uniform props set, i added 3 values to the keyboardShouldPersistTaps:
'never' and 'always' are the same as false and true.
'handled' is the new behavior.

I don't know stance on backward compatibility of the project. The default behavior isn't changed and remains at 'always' (true), but in my opinion, the goal should be to change the default to be the new 'handled' behavior since that feels the most intuitive.

Note that true and false are still valid options.

janicduplessis commented:
I think we should just make `keyboardShouldPersistTaps={false}` your new behaviour instead of changing it to 3 values. It will be less of a breaking change in my opinion, it will change the behaviour a little bit but I don't think there are any use cases where you'd want the current behaviour.

Feels more like a bug fix with `keyboardShouldPersistTaps={false}` than anything else as I'd expect the keyboard to stay opened with touching another input.

cc @mkonicek

MaxLap commented:
I absolutely agree, i just didn't know what was the stance on background compat. If the concensus is that it should be changed straight away, I'll happily change the PR.

janicduplessis commented:
@MaxLap :+1: Let's wait a bit see what others think

satya164 commented:
@janicduplessis I think the current behaviour desirable if your screen filled with touchables. you don't wanna trigger a tap while dismissing the keyboard.

janicduplessis commented:
Does this actually affects touchables? Looks like it only should only affect TextInputs.

satya164 commented:
@janicdupless no it doesn't affect touchables. But when you have a screen filled with touchables, you need to tap somewhere to dismiss the keyboard which was triggered due to an text input without triggering a tap on a touchable.

MaxLap commented:
@janicdupless @satya164 to clarify, the new  behavior basically means if a descendant handles the tap, the ScrollView won't, and therefore won't hide the keyboard. If a descendant button wants to also hide the keyboard, then it can do so itself pretty easily.

MaxLap commented:
Is there a way to request feedback from other people? It's not so much about reviewing the code, first about if this is something you want in RN.

I think this change would be helpful to many people. I admit that some people may want the current swallow-every-tap behavior, which is why I left it there. I also do think the default behavior should the new one, maybe with some releases in between to let people specify which behavior they want.

I didn't change the documentation yet as I wanted feedback first.

janicduplessis commented:
Ping @mkonicek

ericvicenti commented on Libraries/Components/ScrollResponder.js:
```diff
@@ -162,7 +162,14 @@ var ScrollResponderMixin = {
+  scrollResponderHandleStartShouldSetResponder: function(e: Event): boolean {
+    var currentlyFocusedTextInput = TextInputState.currentlyFocusedField();
+
+    if (this.props.keyboardShouldPersistTaps == 'handled' &&
```
please use `===` in cases like this

ericvicenti commented on Libraries/Components/ScrollResponder.js:
```diff
@@ -180,7 +187,9 @@ var ScrollResponderMixin = {
     // First see if we want to eat taps while the keyboard is up
     var currentlyFocusedTextInput = TextInputState.currentlyFocusedField();
-    if (!this.props.keyboardShouldPersistTaps &&
+    var keyboardNeverPersistTaps = !this.props.keyboardShouldPersistTaps ||
```
this could be more consise if you do this on a previous line:
```
const {keyboardShouldPersistTaps} = this.props;
```

janicduplessis commented:
@MaxLap @satya164  I tested this out more in an app and it does affect the interaction with Touchables. If you have a touchable in the screen it will handle the event and thus the keyboard won't be close. Because of this I think we should not change it to the default behavior.

After thinking about it there are use cases for all 3 behaviors
- Persist taps 'always': Never dismiss the keyboard
- Persist taps 'handled': Dismiss the keyboard only when tapping non-touchable views, useful let's say you have a form that include buttons and want to be able to click then without dismissing keyboard beforehand.
- Persist taps 'never': Always dismiss the keyboard even when tapping touchables, useful if you have something like a search bar with a bunch of touchable rows that fill the rest of the screen, you don't want it to trigger the onPress and dismiss the keyboard.

However there is still an issue with the current 'never' behavior (which is I was mostly referring to in my earlier comments) where tapping a different text input while the keyboard is open dismisses the keyboard instead of selecting it. I'll open a new PR to fix this case separately as it doesn't depend on this.

janicduplessis commented:
With that said I'm not the biggest fan of the string enum + boolean prop, what about we add a deprecation warning for the boolean version or create a new prop altogether and deprecate `keyboardShouldPersistTaps`?

MaxLap commented:
I did the changes requested in the review. 

I agree that have both the true/false and the "always"/"never" is probably not the best thing. A deprecation is probably a good idea? Is that the final call?

Making the "never" sometimes still persist the keyboard may make the name of that value confusing.

satya164 commented:
> Persist taps 'never': Always dismiss the keyboard even when tapping touchables, useful if you have something like a search bar with a bunch of touchable rows that fill the rest of the screen, you don't want it to trigger the onPress and dismiss the keyboard.

What happens when I tap non-touchable? Shouldn't the keyboard dismiss then too? Why limit this to touchables only?

> what about we add a deprecation warning for the boolean values

Yeah, we should deprecate the boolean values and show a warning specifying which value to change to.

janicduplessis commented:
@satya164 Maybe that wasnt clear but yea it would dismiss the keyboard when tapping non-touchables too.

satya164 commented:
@janicduplessis great!

MaxLap commented:
Okay, so for the deprecation. Do I make a new prop or I keep using keyboardShouldPersistTaps? If I keep using it, is there a better way of doing the deprecation than this? It works, but the message starts by saying keyboardShouldPersistTaps is deprecated. Is there an alternative / should I code one or that's good enough?

    keyboardShouldPersistTaps: PropTypes.oneOfType(
      [PropTypes.oneOf(['always', 'never', 'handled']),
        deprecatedPropType(PropTypes.bool, 
                           'Values true & false for keyboardShouldPersistTaps ' + 
                           'are deprecated. Use "always" & "never" respectively ' + 
                           'instead or the new "handled".')]),

The resulting message is: keyboardShouldPersistTaps supplied to ScrollView has been deprecated. Values true & false for keyboardShouldPersistTaps are deprecated. Use "always" & "never" respectively instead or the new "handled".

satya164 commented:
> Do I make a new prop or I keep using keyboardShouldPersistTaps

`keyboardShouldPersistTaps` sounds fine to me @janicduplessis?

> Is there an alternative

I think better to check it in `componentDidMount`,

```
const { keyboardShouldPersistTaps } = this.props;
if (keyboardShouldPersistTaps === 'boolean') {
  console.warn(`'keyboardShouldPersistTaps={${keyboardShouldPersistTaps}}' is deprecated. Use 'keyboardShouldPersistTaps={${keyboardShouldPersistTaps ? "always" : "never"}}' instead`);
}
```

ericvicenti commented:
@satya164, fbjs has a warning module:

```
const warning = require('warning');

ericvicenti commented:
@satya164, fbjs has a warning module, which is considered best practice inside RN:

```
const warning = require('warning');
...
warning(
  typeof keyboardShouldPersistTaps !== 'boolean',
  `'keyboardShouldPersistTaps={${keyboardShouldPersistTaps}}' is deprecated. Use 'keyboardShouldPersistTaps=${keyboardShouldPersistTaps ? "always" : "never"}' instead`
);
```

satya164 commented:
@ericvicenti Ah, thanks. didn't know that! So far all deprecation warnings I had added were `console.warn`! Is this new?

ericvicenti commented:
Nope, its been here the whole time, although apparently its not used very widely!

MaxLap commented:
Mixins are a pretty grey area for me (ScrollResponder is a mixin). I don't know yet how it behaves with overriding and such.

Can I just add a function componentDidMount or is there a risk that this overrides / is overriden by something?There is already a componentWillMount defined, so I'm thinking I could do the deprecation warning there instead to avoid this issue. Thoughts?

janicduplessis commented:
@MaxLap `componentWillMount` should work fine.

ericvicenti commented:
Ok, looks great!

MaxLap commented:
I added the deprecation warning and added it to the docs.

Since #10887 (which has been accepted but not yet merged in master) changes the behavior of {false} so that clicking another TextInput doesn't hide the keyboard, the name 'never' will become misleading. I think 'text-input' would be more explicit? At that point, changing from 'always' to 'every' would allow an easier understanding of the prop maybe? keyboard should persist (every / handled / text-input) taps.

Should I do such a change now or wait until #10887 is merged in master?

ericvicenti commented:
I like the idea of doing 'text-input'. When tapping on a disabled input in that mode, I would expect the keyboard to dismiss.

MaxLap commented:
It's pretty hard to read the log of the CI tests and understand what is wrong. Could it be because I didn't update the examples to use the string values yet, the deprecation warnings could be failing the tests?

MaxLap commented:
To clarify your last point @ericvicenti, are your suggesting to add a 4th value 'text-input' or agreeing with me about replacing 'never' with 'text-input' (without changing the behaviour, since that's what  #10887 does)?

janicduplessis commented:
If #10887 works properly with disabled inputs I dont think changing the name is necessary. Touching another input is a bit like if we dismissed and reopened the keyboard immediatly.

ericvicenti commented:
@janicduplessis, are there issues with opening/closing the keyboard quickly on iOS, considering the animation?

I think either name would be fine.

janicduplessis commented:
CI test failures look unrelated, can you try rebasing on master to run the tests again

janicduplessis commented:
@ericvicenti Well it wont actually open / close it, it's just my mental model of it and why I think that `never` is fine even if it doesnt close it on text inputs.

janicduplessis commented on Libraries/Components/ScrollView/ScrollView.js:
```diff
@@ -177,12 +177,18 @@ const ScrollView = React.createClass({
+     *     is up dismisses the keyboard. When this happens, children won't receive the tap.
+     *   - 'always', the keyboard will not dismiss automatically, and the scroll view will not
+     *     catch taps, but children of the scroll view can catch taps.
+     *   - 'handled', they keyboard will not dismiss automatically when the tap was handled by
```
Typo: they -> the

janicduplessis commented on Libraries/Components/ScrollResponder.js:
```diff
@@ -474,6 +486,13 @@ var ScrollResponderMixin = {
    * The `keyboardWillShow` is called before input focus.
    */
   componentWillMount: function() {
+    var keyboardShouldPersistTaps = this.props.keyboardShouldPersistTaps;
```
Nit: lets use destructuring here too for consistency

MaxLap commented:
Did the requested tweaks from the review. Thank you @janicduplessis 

I edited the examples to use string values. I replaced the `true` by `'handled'` since that's probably a more sane behavior for them. I have no idea how to test the examples, is anyone setup to do it relatively painlessly? 

As a side note, am I wrong in thinking that the props and propTypes that refer to keyboardShouldPersistTaps are useless in UIExplorerPage? (only `wrapperProps.keyboardShouldPersistTaps = ...` makes sense) If you agree, I'll remove those useless references.

@janicduplessis's reasoning makes sense to me, so I will stick with 'never'.

janicduplessis commented:
You can run the UIExplorer app from the xcode project in `Examples/UIExplorer` on iOS or with `./gradlew Examples:UIExplorer:android:app:installDebug with a simulator opened on Android. You will also need to start the packager with `./packager/packager.sh`.

Yea, looks like the prop `keyboardShouldPersistTaps` declared in the propTypes is never used we can remove it.

MaxLap commented:
Ok, went easier than expected (I'm on windows with android).

The UIExample works great.

I can't get the Movies example to work. It starts, but it can't seem to do its request to rotten tomatoes: undefined is not an object (evaluating 'responseData.total') at line 136 of SearchScreen.js. Because of this, the ListView is never displayed. I doubt there is an issue with my change since it's a single prop.

janicduplessis commented:
Ok good, as long as uiexplorer works thats fine, anyway your changes don't affect the others.

@ericvicenti Do you want to take care of shipping this since you'll probably want to update fb internal codebase as well to avoid the warning. We should also make sure to communicate this with the release team so they can add something in the release notes about it (not really a breaking change but still).

MaxLap commented:
Anybody willing to make the review on this? Thank you

janicduplessis commented:
Sorry about the delay, I'll ping Eric about it.

ericvicenti commented:
Sorry for the delay on the re-review. This looks great and I'm landing it now!

MaxLap commented:
Thank you everyone
